grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -435,17 +438,78 @@ private static String getResponseBody(List<FieldDescriptor> topLevelFields,
         if (StringUtil.isNullOrEmpty(responseBody)) {
             return null;
         }
-        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+        for (FieldDescriptor fieldDescriptor : topLevelFields) {
             if (fieldDescriptor.getName().equals(responseBody)) {
                 return responseBody;
             }
         }
         return null;
     }
 
+    private static MediaType getMediaTypeFromHttpBody(JsonNode jsonNode) {
+        final String contentType = jsonNode.get("contentType").asText();
+        try {
+            return MediaType.parse(contentType);
+        } catch (IllegalArgumentException e) {
+            logger.warn("Invalid media type in http body content_type {}.", contentType);
+            return MediaType.JSON_UTF_8;
+        }
+    }
+
+    private static Function<HttpData, MediaType> generateMediaTypeDecider(TranscodingSpec spec) {
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpData -> {
+                try (HttpData data = httpData) {
+                    final byte[] array = data.array();
+                    try {
+                        final JsonNode jsonNode = mapper.readValue(array, JsonNode.class);
+                        return getMediaTypeFromHttpBody(jsonNode);
+                    } catch (IOException e) {
+                        logger.warn("Unexpected exception while reading HttpBody {}.", data, e);
+                    }
+
+                    // Default behavior is to return JSON_UTF_8, the default content type for JSON transcoding.
+                    return MediaType.JSON_UTF_8;
```

Comment:
@trustin I think this makes more sense for transcoding perspective. By default responses are json? WDY think?

Reply:
I'm not sure about the others, but my preference is to just set the content type as-is transparently.

So..
- if `non-null`, then set `content-type` transparently
- if `null`, we could just not set the `content-type`. Alternatively, I'm fine with setting a default of `text/plain` which is the default in armeria.

From my understanding, `content-type` isn't really mandatory depending on whether the body `HttpBody` exists. Also, upstream seems to be doing something similar ref: https://github.com/envoyproxy/envoy/blob/94037231eea7150a20b4a07b8c981ae89d1a0a60/source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc#L441

Reply:
What does setting `transparently` mean? Are you implying we shouldn't add any header at all if it is unset? Kinda makes sense.

Reply:
> What does setting transparently mean?

For a little more background, `MediaType` is not really a comprehensive list of all `content-type`s out there.
We pretty much copy over `guava`'s often-used list of `content-type`s manually whenever the following test fails. (we also add custom media types when necessary)

https://github.com/line/armeria/blob/2219de19d510da43b7fc4afcabf730f8a4cca2ce/core/src/test/java/com/linecorp/armeria/common/MediaTypeTest.java#L598

Given that users using `HttpBody` will be sending over arbitrary data, I wasn't sure if it's worth validating the `content-type`.

e.g.
```java
final JsonNode contentType = jsonNode.get("contentType");
if (contentType != null && contentType.isTextual()) {
    builder.set(HttpHeaderNames.CONTENT_TYPE, contentType.textValue());
} else {
    builder.remove(HttpHeaderNames.CONTENT_TYPE);
}
```